src/services/createRouter.spec.ts

```diff
@@ -671,3 +671,42 @@ describe('router.push', () => {
     expect(router.route.state).toMatchObject({ zoo: 123 })
   })
 })
+
+test('global hooks are called correctly', async () => {
+  const router = createRouter(routes, { initialUrl: '/parentA/valueA' })
+
+  const onBeforeRouteEnter = vi.fn()
+  const onBeforeRouteUpdate = vi.fn()
+  const onBeforeRouteLeave = vi.fn()
+  const onAfterRouteEnter = vi.fn()
+  const onAfterRouteUpdate = vi.fn()
+  const onAfterRouteLeave = vi.fn()
+
+  router.onBeforeRouteEnter(onBeforeRouteEnter)
+  router.onAfterRouteEnter(onAfterRouteEnter)
+  router.onBeforeRouteUpdate(onBeforeRouteUpdate)
+  router.onAfterRouteUpdate(onAfterRouteUpdate)
+  router.onBeforeRouteLeave(onBeforeRouteLeave)
+  router.onAfterRouteLeave(onAfterRouteLeave)
+
+  await router.start()
+
+  expect(onBeforeRouteEnter).toHaveBeenCalledTimes(1)
+  expect(onAfterRouteEnter).toHaveBeenCalledTimes(1)
+  
+  await router.push('parentA.childA', { paramA: 'valueA', paramB: 'valueB' })
+  
+  expect(onBeforeRouteEnter).toHaveBeenCalledTimes(2)
```

Comment:
digging into this with you, let me know if you have a breakthrough.

Reply:
also, I prefer to write this with the same expectations after each

```ts
  expect(onBeforeRouteEnter).toHaveBeenCalledTimes(1)
  expect(onBeforeRouteUpdate).toHaveBeenCalledTimes(0)
  expect(onBeforeRouteLeave).toHaveBeenCalledTimes(0)
  expect(onAfterRouteLeave).toHaveBeenCalledTimes(0)
  expect(onAfterRouteUpdate).toHaveBeenCalledTimes(0)
  expect(onAfterRouteEnter).toHaveBeenCalledTimes(1)

  await router.push('parentA.childA', { paramA: 'valueA', paramB: 'valueB' })

  expect(onBeforeRouteEnter).toHaveBeenCalledTimes(2)
  expect(onBeforeRouteUpdate).toHaveBeenCalledTimes(1)
  expect(onBeforeRouteLeave).toHaveBeenCalledTimes(0)
  expect(onAfterRouteLeave).toHaveBeenCalledTimes(0)
  expect(onAfterRouteUpdate).toHaveBeenCalledTimes(1)
  expect(onAfterRouteEnter).toHaveBeenCalledTimes(2)

  await router.push('parentA.childB', { paramA: 'valueB', paramD: 'valueD' })

  expect(onBeforeRouteEnter).toHaveBeenCalledTimes(3)
  expect(onBeforeRouteUpdate).toHaveBeenCalledTimes(2)
  expect(onBeforeRouteLeave).toHaveBeenCalledTimes(1)
  expect(onAfterRouteLeave).toHaveBeenCalledTimes(1)
  expect(onAfterRouteUpdate).toHaveBeenCalledTimes(2)
  expect(onAfterRouteEnter).toHaveBeenCalledTimes(3)
```

Reply:
Good suggestion. I updated both the router test and the plugins test to match this pattern and I updated them to match the current implementation. Updated in https://github.com/kitbagjs/router/pull/397/commits/2bed4556f9ba0b2be33cebca27f9c8851044a8cc

Issue for revisiting hooks https://github.com/kitbagjs/router/issues/402